scripts/lsp_types.lua

```diff
@@ -18,24 +19,117 @@ local function tofile(fname, text)
   end
 end
 
-function M.gen(opt)
-  if vim.uv.fs_stat('./lsp.json') then
-    vim.fn.delete('./lsp.json')
+local function sort_by_method(tbl)
+  local single, client, textD, workspace, others = {}, {}, {}, {}, {}
+  for _, item in ipairs(tbl) do
+    local parts = vim.split(item.method, '/', { trimempty = true })
+    if #parts == 1 then
+      single[#single + 1] = item
+    elseif parts[1] == 'textDocument' then
+      textD[#textD + 1] = item
+    elseif parts[1] == 'client' then
+      client[#client + 1] = item
+    elseif parts[1] == 'workspace' then
+      workspace[#workspace + 1] = item
+    else
+      others[#others + 1] = item
+    end
+  end
+
+  local res = {}
+  for _, item in ipairs({ single, client, textD, workspace, others }) do
+    res = vim.list_extend(res, item)
   end
-  vim.fn.system({
-    'curl',
-    'https://raw.githubusercontent.com/microsoft/lsprotocol/' .. opt.ref .. '/generator/lsp.json',
-    '-o',
-    './lsp.json',
+  return res
+end
+
+local function read_json(opt)
+  local uri = 'https://raw.githubusercontent.com/microsoft/language-server-protocol/gh-pages/_specifications/lsp/'
+    .. opt.version
+    .. '/metaModel/metaModel.json'
+
+  local res = vim.system({ 'curl', uri, '-o', '-' }):wait()
+
+  if res.code ~= 0 then
+    io.write(res.stderr)
+    return
+  end
+  return vim.json.decode(res.stdout)
+end
+
+local function gen_methods(protocol)
+  local output = {
+    '-- Generated by lsp_types.lua, keep at end of file.',
+    '--- LSP method names.',
+    '---',
+    '---@see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#metaModel',
+    'protocol.Methods = {',
+  }
+  local indent = (' '):rep(2)
+
+  for _, item in ipairs(sort_by_method(protocol.requests)) do
+    if item.method then
+      local name = item.method:gsub('/', '_', 3)
+      if item.documentation then
+        local document = vim.split(item.documentation, '\n?\n', { trimempty = true })
+        for _, docstring in ipairs(document) do
+          output[#output + 1] = indent .. '--- ' .. docstring
+        end
+      end
+      output[#output + 1] = indent .. name .. " = '" .. item.method .. "',"
```

Comment:
We should use string:format() to format strings. I don't know why we do this `..` stuff all over the place.